### PR TITLE
Reapply "Teach TypeLayout the Different Types of References" with fixes

### PR DESCRIPTION
**Pull request by gmittert**
This reapplies https://github.com/apple/swift/pull/37116 (reverted here: https://github.com/apple/swift/pull/40142) and fixes several issues with the original Scalar Type Layouts patch:

- AddressOnly types need to be lowered to ScalarKinds that do not load when releasing.
- ClassExistentials now handle the case when the existential is an inline
  aggregation of the class pointer and witness table pointers

The first commit contains the original changes, and the second contains fixes for them.

### CODE REVIEW

File: lib/IRGen/GenExistential.cpp [hunk 1]
```diff
@@ -663,60 +661,52 @@ namespace {
...
+    TypeLayoutEntry *buildTypeLayoutEntry(IRGenModule &IGM,                    \
+                                          SILType T) const override {          \
+      if (Refcounting == ReferenceCounting::Native) {                          \
+        return IGM.typeLayoutCache.getOrCreateScalarEntry(                     \
+            *this, T, ScalarKind::Native##Name##Reference);                    \
+      } else {                                                                 \
+        return IGM.typeLayoutCache.getOrCreateScalarEntry(                     \
+            *this, T, ScalarKind::Unknown##Name##Reference);                   \
+      }                                                                        \
+    }                                                                          \
```
Comment (gmittert) [inline lib/IRGen/GenExistential.cpp:688]: The buildTypeLayoutEntry is the only functional change here, the rest of the macro definition is formatting is sucked in by clang-format.

File: lib/IRGen/GenExistential.cpp [hunk 2]
```diff
@@ -1016,21 +1006,31 @@ class ClassExistentialTypeInfo final
...
-      break;
-    case ReferenceCounting::Unknown:
-      kind = ScalarKind::UnknownReference;
-      break;
-    default:
-      llvm_unreachable("Unsupported refcounting style");
+    // We can't create an objc typeinfo by itself, so don't destructure if we
+    // have one
+    if (Refcounting != ReferenceCounting::Native) {
+      return IGM.typeLayoutCache.getOrCreateTypeInfoBasedEntry(*this, T);
```
Comment (gmittert) [inline lib/IRGen/GenExistential.cpp:1012]: As far as I can find, there is no way to synthesize a Objc Refcounted pointer TypeInfo similar to IGM.getNativeObjectTypeInfo below, so I've set it to fall back to creating a typeinfobased entry here.

Comment (aschwaighofer) [reply lib/IRGen/GenExistential.cpp:1012]: Seems reasonable.

**Review by aschwaighofer — APPROVED** [review]
lgtm

**Comment by gmittert** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@gmittert This breaks the Interpreter/struct_extra_inhabitants.swift test (with an assert-built stdlib and when testing in optimized mode): https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RDA_test-simulator//4151/console

Can you please take a look?

**Comment by eeckstein** [comment]
I temporarily disabled the test in optimize mode: https://github.com/apple/swift/pull/40419

**Comment by aschwaighofer** [comment]
The test failure is an execution time failure. We should revert this until we have a fix instead: https://github.com/apple/swift/pull/40430

**Comment by aschwaighofer** [comment]
The failure can be reproduced with  `ninja check-swift-optimize`

**Comment by aschwaighofer** [comment]
Including the flag ` --test-optimized` in the `build-script` invocation should work.

**Pull request merged**